[DirectX] Add lowering support for `llvm.fsh[l|r].*`

inbelic opened the pull request:
This pr adds support to emulate the `llvm.fshl.*` and `llvm.fshr.*` intrinsics by expanding them, as described [here](https://llvm.org/docs/LangRef.html#llvm-fshl-intrinsic).

Note: the implementation does not create a mask for the first shift as this is automatically done in the DXIL shift equivalents.

Resolves  #165750.

inbelic committed:
[DirectX] Add lowering for fshl

inbelic committed:
[DirectX] Add lowering for fshr

farzonl commented on llvm/test/CodeGen/DirectX/fshl.ll:
```diff
@@ -0,0 +1,82 @@
+; RUN: opt -S -scalarizer -dxil-intrinsic-expansion -mtriple=dxil-pc-shadermodel6.3-library %s | FileCheck %s
+; RUN: opt -S -scalarizer -dxil-intrinsic-expansion -dxil-op-lower -mtriple=dxil-pc-shadermodel6.3-library %s | FileCheck %s
```
In practice we do the scalarizer after intrinsic expansion. Be good to make sure we test it the other way
```suggestion
; RUN: opt -S -dxil-intrinsic-expansion  -scalarizer -mtriple=dxil-pc-shadermodel6.3-library %s | FileCheck %s
; RUN: opt -S -dxil-intrinsic-expansion  -scalarizer -dxil-op-lower -mtriple=dxil-pc-shadermodel6.3-library %s | FileCheck %s
```

farzonl commented on llvm/test/CodeGen/DirectX/fshl.ll:
```diff
@@ -0,0 +1,82 @@
+; CHECK: %[[MASKED_SHIFT0:.*]] = and i64 %[[SHIFT0]], 63
+; CHECK: %[[INVERSE_SHIFT0:.*]] = sub i64 64, %[[MASKED_SHIFT0]]
+; CHECK: %[[RIGHT0:.*]] = lshr i64 %[[B0]], %[[INVERSE_SHIFT0]]
+; CHECK: %[[RES0:.*]] = or i64 %[[LEFT0]], %[[RIGHT0]]
```
can any of these be `CHECK-NEXT`?

farzonl commented on llvm/test/CodeGen/DirectX/fshr.ll:
```diff
@@ -0,0 +1,82 @@
+; RUN: opt -S -scalarizer -dxil-intrinsic-expansion -mtriple=dxil-pc-shadermodel6.3-library %s | FileCheck %s
+; RUN: opt -S -scalarizer -dxil-intrinsic-expansion -dxil-op-lower -mtriple=dxil-pc-shadermodel6.3-library %s | FileCheck %s
```
Same comment here:
```suggestion
; RUN: opt -S -dxil-intrinsic-expansion  -scalarizer -mtriple=dxil-pc-shadermodel6.3-library %s | FileCheck %s
; RUN: opt -S -dxil-intrinsic-expansion  -scalarizer -dxil-op-lower -mtriple=dxil-pc-shadermodel6.3-library %s | FileCheck %s
```

inbelic committed:
review: make stricter with check-next

inbelic committed:
review: use practical order of passes

spall reviewed (changes requested):
From reading the documentation and @lcohedron comment about when Shift == 0 I think this implementation is wrong.

inbelic committed:
review: correct algorithm for edge-case and bits

this fixes two issues with the implementation:

  1. Handles the case when Shift == 0 by separating them into 2
well-defined shifts, following lowerFunnelShiftAsShifts impl
  2. Corrects fshr to extract the least-sig bits instead of most-sig
bits as the algorithm was doing

inbelic committed:
update test cases

inbelic commented on llvm/lib/Target/DirectX/DXILIntrinsicExpansion.cpp:
```diff
@@ -656,6 +659,32 @@ static Value *expandAtan2Intrinsic(CallInst *Orig) {
+  Value *MaskedShift = Builder.CreateAnd(Shift, Mask);
+  Value *InverseShift = Builder.CreateSub(Size, MaskedShift);
+  Value *Right = LeftFunnel ? Builder.CreateLShr(B, InverseShift)
+                            : Builder.CreateShl(A, InverseShift);
```
Good catch, thanks!

We do need to handle the case when it is 0 with a dynamic or static variable, and as pointed out this was missing.

I have updated the implementation to follow the lowering as used in GlobalISel, [here](https://github.com/llvm/llvm-project/blob/71ee84acc4f7c93b9292af90ef5d79dd05687410/llvm/lib/CodeGen/GlobalISel/LegalizerHelper.cpp#L7831). Which accounts for this by creating two shifts of valid amounts.

inbelic commented on llvm/lib/Target/DirectX/DXILIntrinsicExpansion.cpp:
```diff
@@ -656,6 +659,32 @@ static Value *expandAtan2Intrinsic(CallInst *Orig) {
+  Value *Left =
+      LeftFunnel ? Builder.CreateShl(A, Shift) : Builder.CreateLShr(B, Shift);
+
+  Value *MaskedShift = Builder.CreateAnd(Shift, Mask);
```
Added some comments in the new implementation.

spall reviewed (approved):
new algo lgtm
